File: Python/Level 2/땅따먹기 (dp).py
def solution(land):
    dp = [[0,0,0,0] for _ in range(len(land))]
    
    for i in range(4):
        dp[0][i] = land[0][i]
        
    dp[1][0] = land[1][0] + max(dp[0][1],dp[0][2],dp[0][3])
    dp[1][1] = land[1][1] + max(dp[0][0],dp[0][2],dp[0][3])
    dp[1][2] = land[1][2] + max(dp[0][0],dp[0][1],dp[0][3])
    dp[1][3] = land[1][3] + max(dp[0][0],dp[0][1],dp[0][2])
        
    for i in range(2,len(land)):
        dp[i][0] = land[i][0] + max(dp[i-1][1],dp[i-1][2],dp[i-1][3])
        dp[i][1] = land[i][1] + max(dp[i-1][0],dp[i-1][2],dp[i-1][3])
        dp[i][2] = land[i][2] + max(dp[i-1][0],dp[i-1][1],dp[i-1][3])
        dp[i][3] = land[i][3] + max(dp[i-1][0],dp[i-1][1],dp[i-1][2])
        
    return max(dp[len(land)-1][0],dp[len(land)-1][1],dp[len(land)-1][2],dp[len(land)-1][3])

File: Javascript/Level 2/땅따먹기 (dp).js
//function solution(land) {
//    var answer = 0;
//    var dp = new Array(land.length);

//    for (var j = 0; j < land.length; j++) {
//        dp[j] = new Array(4);
//    }

//    dp[0][0] = land[0][0];
//    dp[0][1] = land[0][1];
//    dp[0][2] = land[0][2];
//    dp[0][3] = land[0][3];

//    for (var i = 1; i < land.length; i++) {
//        dp[i][0] = Math.max(dp[i - 1][1], dp[i - 1][2], dp[i - 1][3]) + land[i][0];
//        dp[i][1] = Math.max(dp[i - 1][0], dp[i - 1][2], dp[i - 1][3]) + land[i][1];
//        dp[i][2] = Math.max(dp[i - 1][0], dp[i - 1][1], dp[i - 1][3]) + land[i][2];
//        dp[i][3] = Math.max(dp[i - 1][1], dp[i - 1][2], dp[i - 1][0]) + land[i][3];
//    }

//    answer = Math.max(dp[land.length - 1][0], dp[land.length - 1][1], dp[land.length - 1][2], dp[land.length - 1][3]);

//    return answer;
//}

//// https://programmers.co.kr/learn/courses/30/lessons/12913
//// 2차원 배열을 활용한 dp문제